### PR TITLE
Generator

### PR DESCRIPTION
**Pull request by Ruben-Madelaine**

### CODE REVIEW

**Comment by Manfred-Madelaine-pro** [comment]
ignore the `__pycache__/` directory.

File: validator.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        square = []
+        for j in i:
+            square += j
+        for i in square:
+            if square.count(i) > 1:
+                errors += 1
+    return errors == 0
+
+def valid_solution(board):
+    return valide_line_and_no_zero(board) == valide_column(board)\
```
Comment (Manfred-Madelaine-pro) [inline validator.py]: Can be refactored and shortened

File: generator.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        vals = add_offset(vals, sub)
+        if i > 0 and (i + 1) % sub == 0:
+            vals = add_offset(vals, 1)
+
+    return board
+
+
+def randomizer(board):
+    for i in range(2):
```
Comment (Manfred-Madelaine-pro) [inline generator.py:30]: Careful on that

File: generator.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        random_board = []
+        for i in range(0, len(board), 3):
+            a = board[i : i + 3]
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: Could use the global variable instead of a magic number here (or retrieve it as a function's argument)

File: generator.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+def validation_test(board):
+    test = valid_solution(board)
+    return test
+
+
+def main(num):
+    filled_board = generate_board(sub, size)
+    random_board = mix_loop(filled_board, num)
+
+    print(*random_board, sep="\n")
```
Comment (Manfred-Madelaine-pro) [inline generator.py:48]: clean :+1:

File: generator.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    for i in range(num):
+        board = randomizer(board)
+        board = np.transpose(board).tolist()
+
+    return board
+
+
+def validation_test(board):
+    test = valid_solution(board)
+    return test
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: That's a very short method that we have here

**Review by Manfred-Madelaine-pro — CHANGES_REQUESTED** [review]

File: validator.py (hunk 1 above)
Comment (Ruben-Madelaine) [reply validator.py]: have been updated

File: generator.py (hunk 2 above)
Comment (Ruben-Madelaine) [reply generator.py:30]: fixed

File: generator.py (hunk 3 above)
Comment (Ruben-Madelaine) [reply generator.py]: fixed

File: generator.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+from random import shuffle
+from validator import valid_solution
+
+
+sub = 3
+size = sub ** 2
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: Use capital case

File: generator.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        if i > 0 and (i + 1) % sub == 0:
+            vals = add_offset(vals, 1)
+
+    return board
+
+
+def randomizer(board,sub):
+    random_board = []
+    for i in range(0, len(board), sub):
+        a = board[i : i + sub]
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: Rename:
* `a` to `group_of_rows`
and 
* `sub` to `SUB_SIZE`

File: generator.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
```
Comment (Manfred-Madelaine-pro) [inline generator.py:1]: Do we need more information in README about the `numpy` package?

File: generator.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+def randomizer(board,sub):
+    random_board = []
+    for i in range(0, len(board), sub):
+        a = board[i : i + sub]
+        shuffle(a)
+        random_board += a
+    return random_board
+
+
+def mix_loop(board, num):
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: Maybe rename `num` to a more specific name like `iter` for iterations

File: generator.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    for i in range(num):
+        board = randomizer(board,sub)
```
Comment (Manfred-Madelaine-pro) [inline generator.py]: Apply formater

Comment (Manfred-Madelaine-pro) [inline generator.py]: Don't know if using `sub` global variable here is the best way to do it

**Review by Manfred-Madelaine-pro — CHANGES_REQUESTED** [review]

**Review by Manfred-Madelaine-pro — APPROVED** [review]

**Pull request merged**